### ISSUE TITLE
CVL  much too high

### ISSUE THREAD

**Issue by MatKra89** [Louisvdw/dbus-serialbattery#1094]
### Describe the bug

There is a massive problem with version v1.3.20240705. 
The CVL is wrong and almost shot my cells. 
Both the latest 1.3.20240705 and the nightly build etc. are affected. 
Apparently the software reads CVL from the BMS (max cell voltage). 
In Serialbattery I set 3.45V (bulk) and 3.35V (float) as the maximum voltage, but serialbattery still takes the full 3.65V from the BMS. When the cells reach this voltage, the value jumps back and forth. This means that it charges briefly at full power, then overshoots to 3.6V/cell and then immediately discharges again. I'm now reinstalling an old version, the problem didn't exist there. 
Or am I doing something wrong?

I have a 16s system and that would be a total of 55.2V CVL at 3.45V. With float it would still be 53.6V. 
Here you can see the 58.4V CVL....
![grafik](https://github.com/Louisvdw/dbus-serialbattery/assets/101696060/8a5a6114-f1aa-4a39-bf01-521bd68551af)

### How to reproduce

cvl that is too high is constantly displayed...

### Expected behavior

What is expected is a cvl similar to the "old" version, i.e. maximum cell voltage * number.... in my case 3.45V * 16 = 55.2V

### Driver version

v1.3.20240705

### Venus OS device type

Raspberry Pi 3

### Venus OS version

3.33

### BMS type

JKBMS / Heltec BMS

### Cell count

16

### Battery count

3

### Connection type

Bluetooth

### Config file

```ini
[DEFAULT]

; --------- Set logging level ---------
LOGGING = INFO

; --------- Battery Current limits ---------
MAX_BATTERY_CHARGE_CURRENT = 80.0
MAX_BATTERY_DISCHARGE_CURRENT = 80.0

; --------- Cell Voltages ---------
MIN_CELL_VOLTAGE   = 3.000
MAX_CELL_VOLTAGE   = 3.450
FLOAT_CELL_VOLTAGE = 3.350

; --------- Bluetooth BMS ---------
BLUETOOTH_BMS = Jkbms_Ble C8:47:80:03:8A:2C, Jkbms_Ble C8:47:80:03:82:4A, Jkbms_Ble C8:47:80:03:89:20,

; --------- Bluetooth use USB ---------
BLUETOOTH_USE_USB = False

; --------- BMS disconnect behaviour ---------
BLOCK_ON_DISCONNECT = True

; --------- Charge mode ---------
LINEAR_LIMITATION_ENABLE = True
LINEAR_RECALCULATION_EVERY = 60
LINEAR_RECALCULATION_ON_PERC_CHANGE = 5

; --------- Charge Voltage limitation (affecting CVL) ---------
CVCM_ENABLE = True
CELL_VOLTAGE_DIFF_KEEP_MAX_VOLTAGE_UNTIL        = 0.010
CELL_VOLTAGE_DIFF_KEEP_MAX_VOLTAGE_TIME_RESTART = 0.015
CELL_VOLTAGE_DIFF_TO_RESET_VOLTAGE_LIMIT        = 0.080
MAX_VOLTAGE_TIME_SEC = 600
SOC_LEVEL_TO_RESET_VOLTAGE_LIMIT = 80

; --------- SOC calculation ---------
SOC_CALCULATION   = False

; --------- Cell Voltage Current limitation (affecting CCL/DCL) ---------
CCCM_CV_ENABLE = True
DCCM_CV_ENABLE = True
CELL_VOLTAGES_WHILE_CHARGING      = 3.55, 3.45, 3.42, 3.38, 3.30
MAX_CHARGE_CURRENT_CV_FRACTION    =    0, 0.07, 0.7, 1, 1
CELL_VOLTAGES_WHILE_DISCHARGING   = 2.70, 2.80, 2.90, 3.10
MAX_DISCHARGE_CURRENT_CV_FRACTION =    0,  0.1,  0.5,    1

; --------- Cell Voltage limitation (affecting CVL) ---------
CVL_ICONTROLLER_MODE   = False
CVL_ICONTROLLER_FACTOR = 0.2

; --------- Temperature limitation (affecting CCL/DCL) ---------
CCCM_T_ENABLE = True
DCCM_T_ENABLE = True
TEMPERATURES_WHILE_CHARGING      =    0,    2,    5,   10,   15,   20,   35,   40,   55
MAX_CHARGE_CURRENT_T_FRACTION    = 0.00, 0.10, 0.20, 0.50, 1.00, 1.00, 1.00, 0.40, 0.00
TEMPERATURES_WHILE_DISCHARGING   =  -20,    0,    5,   10,   15,   45,   55
MAX_DISCHARGE_CURRENT_T_FRACTION = 0.00, 0.20, 0.50, 0.80, 1.00, 1.00, 0.00

; --------- Additional settings ---------
;BMS_TYPE = Jkbms
AUTO_RESET_SOC = False
PUBLISH_CONFIG_VALUES = True
BATTERY_CELL_DATA_FORMAT = 0
MIDPOINT_ENABLE = False
TEMP_BATTERY = 1
TEMP_1_NAME = Temp Akku 1
TEMP_2_NAME = Temp Akku 2
TEMP_3_NAME = Temp Akku 3
TEMP_4_NAME = Temp Akku 4

; --------- BMS specific settings ---------
JKBMS_CAN_CELL_COUNT = 16
```

### Relevant log output

```shell
2024-07-08 16:54:31.316384500 *** CCGX booted (0) ***

==> /data/log/dbus-blebattery.0/current <==
2024-07-08 16:55:26.041405500
2024-07-08 16:55:26.041409500 INFO:Bluetooth details
2024-07-08 16:55:26.413926500 Attempting to disconnect from C8:47:80:03:8A:2C
2024-07-08 16:55:26.413934500 Successful disconnected

==> /data/log/dbus-blebattery.1/current <==
2024-07-08 16:55:26.118110500
2024-07-08 16:55:26.118116500 INFO:Bluetooth details
2024-07-08 16:55:26.438130500 Attempting to disconnect from C8:47:80:03:82:4A
2024-07-08 16:55:26.438136500 Successful disconnected

==> /data/log/dbus-blebattery.2/current <==
2024-07-08 16:55:26.114153500
2024-07-08 16:55:26.114157500 INFO:Bluetooth details
2024-07-08 16:55:26.543551500 Attempting to disconnect from C8:47:80:03:89:20
2024-07-08 16:55:26.543557500 Successful disconnected

==> /data/log/dbus-blebattery.0/current <==
2024-07-08 16:55:31.914927500 Device C8:47:80:03:8A:2C (public)
2024-07-08 16:55:31.914935500   Alias: E&H 02
2024-07-08 16:55:31.914937500   Paired: no
2024-07-08 16:55:31.914939500   Trusted: no
2024-07-08 16:55:31.914941500   Blocked: no
2024-07-08 16:55:31.914942500   Connected: no
2024-07-08 16:55:31.914944500   LegacyPairing: no
2024-07-08 16:55:31.914946500   UUID: Device Information        (0000180a-0000-1000-8000-00805f9b34fb)
2024-07-08 16:55:31.914950500   RSSI: -52
2024-07-08 16:55:31.917633500

==> /data/log/dbus-blebattery.1/current <==
2024-07-08 16:55:31.909116500 Device C8:47:80:03:82:4A (public)
2024-07-08 16:55:31.909123500   Alias: E&H 01
2024-07-08 16:55:31.909125500   Paired: no
2024-07-08 16:55:31.909126500   Trusted: no
2024-07-08 16:55:31.909128500   Blocked: no
2024-07-08 16:55:31.909130500   Connected: no
2024-07-08 16:55:31.909132500   LegacyPairing: no
2024-07-08 16:55:31.909134500   UUID: Device Information        (0000180a-0000-1000-8000-00805f9b34fb)
2024-07-08 16:55:31.909138500   RSSI: -57
2024-07-08 16:55:31.913529500

==> /data/log/dbus-blebattery.2/current <==
2024-07-08 16:55:31.979882500 Device C8:47:80:03:89:20 (public)
2024-07-08 16:55:31.979890500   Alias: E&H 03
2024-07-08 16:55:31.979892500   Paired: no
2024-07-08 16:55:31.979894500   Trusted: no
2024-07-08 16:55:31.979896500   Blocked: no
2024-07-08 16:55:31.979897500   Connected: no
2024-07-08 16:55:31.979899500   LegacyPairing: no
2024-07-08 16:55:31.979901500   UUID: Device Information        (0000180a-0000-1000-8000-00805f9b34fb)
2024-07-08 16:55:31.979905500   RSSI: -60
2024-07-08 16:55:31.981098500

==> /data/log/dbus-blebattery.0/current <==
2024-07-08 16:55:39.838524500 INFO:SerialBattery:
2024-07-08 16:55:39.839310500 INFO:SerialBattery:Starting dbus-serialbattery
2024-07-08 16:55:39.841036500 INFO:SerialBattery:Venus OS v3.33
2024-07-08 16:55:39.841611500 INFO:SerialBattery:dbus-serialbattery v1.3.20240705

==> /data/log/dbus-blebattery.1/current <==
2024-07-08 16:55:39.820028500 INFO:SerialBattery:
2024-07-08 16:55:39.820235500 INFO:SerialBattery:Starting dbus-serialbattery
2024-07-08 16:55:39.840325500 INFO:SerialBattery:Venus OS v3.33
2024-07-08 16:55:39.843495500 INFO:SerialBattery:dbus-serialbattery v1.3.20240705

==> /data/log/dbus-blebattery.2/current <==
2024-07-08 16:55:39.873471500 INFO:SerialBattery:
2024-07-08 16:55:39.874267500 INFO:SerialBattery:Starting dbus-serialbattery
2024-07-08 16:55:39.875956500 INFO:SerialBattery:Venus OS v3.33
2024-07-08 16:55:39.876535500 INFO:SerialBattery:dbus-serialbattery v1.3.20240705

==> /data/log/dbus-blebattery.0/current <==
2024-07-08 16:55:57.073464500 INFO:SerialBattery:Init of Jkbms_Ble at C8:47:80:03:8A:2C
2024-07-08 16:55:57.074408500 INFO:SerialBattery:Test of Jkbms_Ble at C8:47:80:03:8A:2C

==> /data/log/dbus-blebattery.1/current <==
2024-07-08 16:55:57.110468500 INFO:SerialBattery:Init of Jkbms_Ble at C8:47:80:03:82:4A
2024-07-08 16:55:57.111283500 INFO:SerialBattery:Test of Jkbms_Ble at C8:47:80:03:82:4A

==> /data/log/dbus-blebattery.2/current <==
2024-07-08 16:55:57.057937500 INFO:SerialBattery:Init of Jkbms_Ble at C8:47:80:03:89:20
2024-07-08 16:55:57.058859500 INFO:SerialBattery:Test of Jkbms_Ble at C8:47:80:03:89:20

==> /data/log/dbus-blebattery.0/current <==
2024-07-08 16:56:03.668248500 INFO:SerialBattery:BAT: JKBMS 11.XW 16 cells (20240408)
2024-07-08 16:56:03.668254500 INFO:SerialBattery:Connection established to Jkbms_Ble

==> /data/log/dbus-blebattery.1/current <==
2024-07-08 16:56:03.697131500 INFO:SerialBattery:BAT: JKBMS 11.XW 16 cells (20240408)
2024-07-08 16:56:03.698739500 INFO:SerialBattery:Connection established to Jkbms_Ble

==> /data/log/dbus-blebattery.2/current <==
2024-07-08 16:56:03.632911500 INFO:SerialBattery:BAT: JKBMS 11.XW 16 cells (20240408)
2024-07-08 16:56:03.634127500 INFO:SerialBattery:Connection established to Jkbms_Ble

==> /data/log/dbus-blebattery.0/current <==
2024-07-08 16:56:06.670696500 INFO:SerialBattery:Found existing battery with DeviceInstance = 3

==> /data/log/dbus-blebattery.1/current <==
2024-07-08 16:56:06.736391500 INFO:SerialBattery:Found existing battery with DeviceInstance = 1

==> /data/log/dbus-blebattery.2/current <==
2024-07-08 16:56:06.718441500 INFO:SerialBattery:Found existing battery with DeviceInstance = 2

==> /data/log/dbus-blebattery.0/current <==
2024-07-08 16:56:07.166763500 INFO:SerialBattery:DeviceInstance = 3
2024-07-08 16:56:07.168158500 INFO:SerialBattery:PID file created successfully: /var/tmp/dbus-serialbattery_3.pid
2024-07-08 16:56:07.168654500 INFO:SerialBattery:Used DeviceInstances = ['2', '1', '3']
2024-07-08 16:56:07.169186500 INFO:SerialBattery:com.victronenergy.battery.ble_c84780038a2c
2024-07-08 16:56:07.170868500 INFO:SerialBattery:BAT: JKBMS 11.XW 16 cells (20240408)
2024-07-08 16:56:07.214254500 INFO:SerialBattery:publish config values = True
2024-07-08 16:56:07.257467500 INFO:SerialBattery:Battery Jkbms_Ble connected to dbus from ble_c84780038a2c
2024-07-08 16:56:07.258670500 INFO:SerialBattery:========== Settings ==========
2024-07-08 16:56:07.259580500 INFO:SerialBattery:> Connection voltage: 54.01 V | Current: -1.5 A | SoC: 100%
2024-07-08 16:56:07.260150500 INFO:SerialBattery:> Cell count: 16 | Cells populated: 32
2024-07-08 16:56:07.260803500 INFO:SerialBattery:> LINEAR LIMITATION ENABLE: True
2024-07-08 16:56:07.261369500 INFO:SerialBattery:> MIN CELL VOLTAGE: 3.000 V | MAX CELL VOLTAGE: 3.450 V| FLOAT CELL VOLTAGE: 3.350 V
2024-07-08 16:56:07.261899500 INFO:SerialBattery:> MAX BATTERY CHARGE CURRENT: 80.0 A | MAX BATTERY DISCHARGE CURRENT: 80.0 A
2024-07-08 16:56:07.262420500 INFO:SerialBattery:> MAX BATTERY CHARGE CURRENT: 130.0 A | MAX BATTERY DISCHARGE CURRENT: 130.0 A (read from BMS)
2024-07-08 16:56:07.262907500 INFO:SerialBattery:> CVCM:     True
2024-07-08 16:56:07.263464500 INFO:SerialBattery:> CCCM CV:  True  | DCCM CV:  True
2024-07-08 16:56:07.269426500 INFO:SerialBattery:> CCCM T:   True  | DCCM T:   True
2024-07-08 16:56:07.269434500 INFO:SerialBattery:> CCCM SOC: False | DCCM SOC: False
2024-07-08 16:56:07.271204500 INFO:SerialBattery:> CHARGE FET: True | DISCHARGE FET: True | BALANCE FET: True
2024-07-08 16:56:07.271493500 INFO:SerialBattery:Serial Number/Unique Identifier: 3071245305

==> /data/log/dbus-blebattery.1/current <==
2024-07-08 16:56:07.198767500 INFO:SerialBattery:DeviceInstance = 1
2024-07-08 16:56:07.224017500 INFO:SerialBattery:PID file created successfully: /var/tmp/dbus-serialbattery_1.pid
2024-07-08 16:56:07.224951500 INFO:SerialBattery:Used DeviceInstances = ['2', '1', '3']
2024-07-08 16:56:07.226118500 INFO:SerialBattery:com.victronenergy.battery.ble_c8478003824a
2024-07-08 16:56:07.228548500 INFO:SerialBattery:BAT: JKBMS 11.XW 16 cells (20240408)
2024-07-08 16:56:07.267297500 INFO:SerialBattery:publish config values = True
2024-07-08 16:56:07.303810500 INFO:SerialBattery:Battery Jkbms_Ble connected to dbus from ble_c8478003824a
2024-07-08 16:56:07.304502500 INFO:SerialBattery:========== Settings ==========
2024-07-08 16:56:07.305202500 INFO:SerialBattery:> Connection voltage: 54.01 V | Current: -1.5 A | SoC: 100%
2024-07-08 16:56:07.306166500 INFO:SerialBattery:> Cell count: 16 | Cells populated: 32
2024-07-08 16:56:07.308085500 INFO:SerialBattery:> LINEAR LIMITATION ENABLE: True
2024-07-08 16:56:07.308915500 INFO:SerialBattery:> MIN CELL VOLTAGE: 3.000 V | MAX CELL VOLTAGE: 3.450 V| FLOAT CELL VOLTAGE: 3.350 V
2024-07-08 16:56:07.309498500 INFO:SerialBattery:> MAX BATTERY CHARGE CURRENT: 80.0 A | MAX BATTERY DISCHARGE CURRENT: 80.0 A
2024-07-08 16:56:07.312809500 INFO:SerialBattery:> MAX BATTERY CHARGE CURRENT: 130.0 A | MAX BATTERY DISCHARGE CURRENT: 130.0 A (read from BMS)
2024-07-08 16:56:07.313574500 INFO:SerialBattery:> CVCM:     True
2024-07-08 16:56:07.314246500 INFO:SerialBattery:> CCCM CV:  True  | DCCM CV:  True
2024-07-08 16:56:07.314916500 INFO:SerialBattery:> CCCM T:   True  | DCCM T:   True
2024-07-08 16:56:07.315503500 INFO:SerialBattery:> CCCM SOC: False | DCCM SOC: False
2024-07-08 16:56:07.316176500 INFO:SerialBattery:> CHARGE FET: True | DISCHARGE FET: True | BALANCE FET: True
2024-07-08 16:56:07.316689500 INFO:SerialBattery:Serial Number/Unique Identifier: 3071245268

==> /data/log/dbus-blebattery.2/current <==
2024-07-08 16:56:07.157889500 INFO:SerialBattery:DeviceInstance = 2
2024-07-08 16:56:07.160688500 INFO:SerialBattery:PID file created successfully: /var/tmp/dbus-serialbattery_2.pid
2024-07-08 16:56:07.162338500 INFO:SerialBattery:Used DeviceInstances = ['2', '1', '3']
2024-07-08 16:56:07.163182500 INFO:SerialBattery:com.victronenergy.battery.ble_c84780038920
2024-07-08 16:56:07.164475500 INFO:SerialBattery:BAT: JKBMS 11.XW 16 cells (20240408)
2024-07-08 16:56:07.188155500 INFO:SerialBattery:publish config values = True
2024-07-08 16:56:07.233599500 INFO:SerialBattery:Battery Jkbms_Ble connected to dbus from ble_c84780038920
2024-07-08 16:56:07.234399500 INFO:SerialBattery:========== Settings ==========
2024-07-08 16:56:07.235163500 INFO:SerialBattery:> Connection voltage: 54.01 V | Current: -1.7 A | SoC: 100%
2024-07-08 16:56:07.235814500 INFO:SerialBattery:> Cell count: 16 | Cells populated: 32
2024-07-08 16:56:07.236448500 INFO:SerialBattery:> LINEAR LIMITATION ENABLE: True
2024-07-08 16:56:07.237133500 INFO:SerialBattery:> MIN CELL VOLTAGE: 3.000 V | MAX CELL VOLTAGE: 3.450 V| FLOAT CELL VOLTAGE: 3.350 V
2024-07-08 16:56:07.237859500 INFO:SerialBattery:> MAX BATTERY CHARGE CURRENT: 80.0 A | MAX BATTERY DISCHARGE CURRENT: 80.0 A
2024-07-08 16:56:07.238581500 INFO:SerialBattery:> MAX BATTERY CHARGE CURRENT: 130.0 A | MAX BATTERY DISCHARGE CURRENT: 130.0 A (read from BMS)
2024-07-08 16:56:07.239219500 INFO:SerialBattery:> CVCM:     True
2024-07-08 16:56:07.239871500 INFO:SerialBattery:> CCCM CV:  True  | DCCM CV:  True
2024-07-08 16:56:07.240560500 INFO:SerialBattery:> CCCM T:   True  | DCCM T:   True
2024-07-08 16:56:07.241161500 INFO:SerialBattery:> CCCM SOC: False | DCCM SOC: False
2024-07-08 16:56:07.241817500 INFO:SerialBattery:> CHARGE FET: True | DISCHARGE FET: True | BALANCE FET: True
2024-07-08 16:56:07.248243500 INFO:SerialBattery:Serial Number/Unique Identifier: 3071245114
```

**Comment by mr-manuel** [comment]
What I see in your log is:

```
2024-07-08 16:56:07.235814500 INFO:SerialBattery:> Cell count: 16 | Cells populated: 32
```

`Cells populated: 32` is wrong. Can you check this on an old version, if you have that also? I have many systems with JKBMS, but none gives me that issue.

Please try also the latest nightly and enable `GUI_PARAMETERS_SHOW_ADDITIONAL_INFO = True`. The reinstall the driver, go to the parameters page and post the screenshots of the items below DCL.

**Comment by MatKra89** [comment]
Version v1.2.20240408 is currently running again and here too it says Cell count: 16 | Cells populated: 32

Now that the “old” version is running again, I have reinstalled the current v1.3.20240705. 
very interesting, because now CVL is suddenly correct??
I changed several times before and it was always wrong. Maybe there was something else cached? 
Tomorrow I'll test the behavior of the nightly build when the battery is full. CVL currently fits.....
![grafik](https://github.com/Louisvdw/dbus-serialbattery/assets/101696060/292d5f1c-3b0a-47d9-b8c8-f4779732cfca)

**Comment by mr-manuel** [comment]
Make sure to install the nightly, since it contains a fix in combination with `GUI_PARAMETERS_SHOW_ADDITIONAL_INFO`, else nothing other changed. See https://github.com/mr-manuel/venus-os_dbus-serialbattery/issues/48.

**Comment by MatKra89** [comment]
The nightly is now running with the fix, so far so good. No abnormalities so far. 👍
I'll test this for a few days...

First of all, thank you very much for the quick help =)

**Comment by MatKra89** [comment]
The nightly build has been running since yesterday evening in conjunction with GUI_PARAMETERS_SHOW_ADDITIONAL_INFO = True. 
Unfortunately the bulk voltage is still much too high, 55.2V would be correct.
![grafik](https://github.com/Louisvdw/dbus-serialbattery/assets/101696060/8c5c209c-7b20-4ea9-88da-209f721026d5)

![grafik](https://github.com/Louisvdw/dbus-serialbattery/assets/101696060/5af18d3e-3d5d-4df8-8ca2-326ea565747c)

![grafik](https://github.com/Louisvdw/dbus-serialbattery/assets/101696060/f6dcfa6f-51b4-4508-bf42-1432cd23910d)

![grafik](https://github.com/Louisvdw/dbus-serialbattery/assets/101696060/9d075abd-7dee-40ab-8133-60960a4ce305)

![grafik](https://github.com/Louisvdw/dbus-serialbattery/assets/101696060/3c91ca37-483e-4a5c-bbf8-e84b88efe59f)



I have a guess... my battery was almost full when I updated serialbattery. Since then it hasn't seen the 80% trigger on bulk. Is it possible that the correct CVL value is written when the trigger threshold falls below 80% and the CVL value is only incorrect immediately after installation?

**Comment by mr-manuel** [comment]
Could you try to change
https://github.com/Louisvdw/dbus-serialbattery/blob/fd0fc1a193427df6ca0fadd141aa57826f761819/etc/dbus-serialbattery/config.default.ini#L46
to 3.50, so we see, if the value is comming from there

**Comment by mr-manuel** [comment]
Can you also write me on Discord? You find the link by clicking on my profile.

**Comment by MatKra89** [comment]
Today one of the three battery blocks went from float to bulk. Unfortunately, again the CVL is too high.
I set SOC_RESET_VOLTAGE = 3,500 in the config and rebooted. Unfortunately also without success.

I don't use Discord, I could offer email or WhatsApp.
Both public because YouTubers ^^

**Comment by mr-manuel** [comment]
I would need access to your system, please send me your e-Mail, then I contact you.

**Comment by MatKra89** [comment]
kramer-matthias@gmx.de

**Comment by mr-manuel** [comment]
I deleted your comment for your privacy. I write you.

**Comment by mr-manuel** [comment]
Did you get my e-Mail?

**Comment by MatKra89** [comment]
***@***.*** Am 10.07.24, 18:44 schrieb Manuel ***@***.***>:

   
  I would need access to your system, please send me your e-Mail, then I contact you.

**Comment by mr-manuel** [comment]
The bug was fixed with `v1.4.20240712dev`.
